File: java/MergeSort.java
/*
 * Merge Sort Implementation
 */

public class MergeSort {
	
	private int[] array;
	private int length;
	private int[] tempArray;

	public static void main(String[] args) //throws InterruptedException
		{
		int[] itemsToSort = {45,23,11,89,77,98,4,28,65,43};
		System.out.println("Pre-Sort");
		for (int i: itemsToSort){
			System.out.print(i);
			System.out.print(" ");
		}
		System.out.print("\n");
		MergeSort ms = new MergeSort();
		ms.sort(itemsToSort);
		
		System.out.println("PrintOut:");
		for (int i: itemsToSort){
			System.out.print(i);
			System.out.print(" ");
		}
		}

	public void sort( int arr[] ) //throws InterruptedException
		{
		this.array = arr;
		this.length = arr.length;
		this.tempArray = new int[this.length];

		sort(0, this.length -1);
		}

	private void sort ( int lowerBounds, int upperBounds) //throws InterruptedException
		{
		//Thread.sleep(500);
		if (lowerBounds < upperBounds)
			{
			//if ( (upperBounds - lowerBounds) <= 1) return;
			int center = (upperBounds - lowerBounds)/2 + lowerBounds;
			System.out.println("Lower: " + lowerBounds + "\tCenter: " + center + "\tUpper: " + upperBounds);
			sort(lowerBounds, center);
			sort(center + 1, upperBounds);
			merge(lowerBounds,center, upperBounds);
			}
		}

	private void merge ( int lowerBounds, int center, int upperBounds)
		{
		for (int i = lowerBounds; i <= upperBounds; i++)
			{
			tempArray[i] = array[i];
			}

		int i = lowerBounds;
		int j = center + 1;
		int k = lowerBounds;

		while(i <= center && j <= upperBounds)
			{
			if (tempArray[i] <= tempArray[j])
				{
				array[k] = tempArray[i];
				i++;
				}
			else
				{
				array[k] = tempArray[j];
				j++;
				}
			k++;
			}
		while ( i <= center )
			{
			array[k] = tempArray[i];
			k++;
			i++;
			}
		}

}

File: javascript/Merge_Sort.js


class MergeSort {
    constructor(list){
        this.list = list;
        this.tempArray = new Array(this.list.length);
        this.sort(0, this.list.length - 1)
        return this.list;
    }

    sort(lowerIdx, upperIdx){
        if (lowerIdx < upperIdx){
            let centerIdx = Math.floor((upperIdx - lowerIdx) / 2) + lowerIdx;
            this.sort(lowerIdx, centerIdx);
            this.sort(centerIdx + 1, upperIdx);
            console.log("Sort: ", lowerIdx, centerIdx, upperIdx)
            this.merge(lowerIdx, centerIdx, upperIdx);
        }
    }
    merge(lowerIdx, centerIdx, upperIdx){
        for (let i = lowerIdx; i <= upperIdx; i++){
            this.tempArray[i] = this.list[i];
        }
        console.log("Temp Array: ", this.tempArray);
        let i = lowerIdx, j = centerIdx + 1, k = lowerIdx;
        while ( i <= centerIdx && j <= upperIdx) {
            if ( this.tempArray[i] <= this.tempArray[j] ) {
                this.list[k] = this.tempArray[i]
                i ++;
            } else {
                this.list[k] = this.tempArray[j]
                j ++;
            }
            k++;
        }
        
        while ( i <= centerIdx ) {
            this.list[k] = this.tempArray[i];
            i ++;
            k ++;
        }
        
    }
}



function main(){
    let array = [];
    for (let i = 0; i < 10; i ++) {
        array.push(Math.floor(Math.random() * 100 + 1))
    }
    console.log(array);
    console.log("Results: ", new MergeSort(array))
}
main();